daemon/serve.cpp

```diff
@@ -240,12 +240,12 @@ int handle_connection(const string &basedir, CompileJob *job,
 
             string output_dir, relative_file_path;
             if (!file_dir.empty() && file_dir[0] == '/') { // output dir is absolute, convert to relative
-                relative_file_path = get_relative_path(get_canonicalized_path(job_output_file), get_canonicalized_path(job_working_dir));
-                output_dir = tmp_path + get_canonicalized_path(file_dir);
+                relative_file_path = get_relative_path(get_abspath(job_output_file), get_abspath(job_working_dir));
+                output_dir = tmp_path + get_abspath(file_dir);
             }
             else { // output file is already relative, canonicalize in relation to working dir
-                string canonicalized_dir = get_canonicalized_path(job_working_dir + '/' + file_dir);
-                relative_file_path = get_relative_path(canonicalized_dir + '/' + file_name, get_canonicalized_path(job_working_dir));
+                string canonicalized_dir = get_abspath(job_working_dir + '/' + file_dir);
+                relative_file_path = get_relative_path(canonicalized_dir + '/' + file_name, get_abspath(job_working_dir));
                 output_dir = tmp_path + canonicalized_dir;
```

Comment:
There is a subtle bug here -

Changing the implementation to use `realpath` _could_ break the calculation of the working path and output file path that is used for determining where to build from and what relative path to use to the output file.

Consider the following: the host machine is attempting to build a file from one working directory, targeting the output to another directory. They send the job to icecream. The job ends up on a machine where either that working directory or the target output directory *exist as symlinks*. This code will now resolve that path to the actual path on the *build* machine, and append that to the `/tmp/` chroot folder, breaking the relative path structure needed for correct `.dwo` file paths. This actually leaks information about the build machine and breaks the resulting dwarf object information.

We truly do want to purely canonicalize the given path, *not* resolve the symlinks as well.

Reply:
> The job ends up on a machine where either that working directory or the target output directory exist as symlinks.

Is there any (non-trivial) path handling that occurs on the remote end?  I need to confirm, but aren't the remote commands handled inside a (clean) chroot?

Having said that, if we don't need to resolve symlinks then ignoring the fs is going to be faster.  I will update the PR once I find time to update and retest.

Reply:
This part (where it's determining the paths) is done before it chroots. It needs to figure out which directories to create inside that chroot. 

Resolving symlinks is not something this part of the code should do since it's trying to mimic the initiatior's filesystem--it shouldn't be resolving things on the remote system.

Reply:
Updated the PR, removing symlink resolution.